src/box/xlog.c

```diff
@@ -420,10 +422,11 @@ xdir_index_file(struct xdir *dir, int64_t signature)
 	}
 
 	/*
-	 * Append the clock describing the file to the
-	 * directory index.
+	 * Append the clock describing the file to the directory index.
+	 * If retention feature is available (EE), then additional memory is
+	 * allocated. Time, when xlog protection should stop is saved there.
 	 */
-	struct vclock *vclock = (struct vclock *) malloc(sizeof(*vclock));
+	struct vclock *vclock = retention_vclock_create();
```

Comment:
It seems you may actually use `xmalloc` in `retention_vclock_create()`.
Looks like all the code which checks `xdir_scan` errors (it calls `xdir_index_file`), makes tarantool exit with an error.
So I see nothing bad in panicking in case of malloc failure.

Reply:
It was `xlog_cursor_close`, which was called, when `vclock` allocation failed in `xdir_index_file`. Will it be closed correctly in case of panic?

Anyway, I've used `xmalloc` for now.

Reply:
> Will it be closed correctly in case of panic?

Yes. Besides, there are other places where we panic, and none of the open files are closed in these cases.